### DISCUSSION TITLE
Wrapper SingleInstance Codeunit for Setup Tables

### DISCUSSION THREAD

**Discussion by StefanMaron** [microsoft/alguidelines#91] [BC Patterns]
Some setup tables may have massive access, and therefore also read rates.

A solution for this can be to wrap the reads through a single instance codeunit.  
This codeunit holds the setup table as a global record variable and saves weather it has already been read.  
When the setup table is updated, a procedure in the wrapper codeunit is called to update the saved record.

Maybe the RecordRef.IsDirty() method can be used for this:  
https://docs.microsoft.com/de-de/dynamics365/business-central/dev-itpro/developer/methods-auto/recordref/recordref-isdirty-method

**Comment by waldo1001** [comment]
Interested!

I personally never cared about setup-records being read regularly.  The reason is the NST cache .. since it's in cache, there is not a lot to optimize, is there?

**Reply by StefanMaron** [reply]
Ah NST cache, you are right. I will try to find the time to test the performance gain with a single instance codeunit

**Reply by NKarolak** [reply]
My principle: don't rely on some cache or other technical features, but on your code 🤷 (however, in test apps I'm kinda lazy)

**Reply by waldo1001** [reply]
Fair enough.  I do "trust" the platform in that perspective.

I do implement caching architecture when there is no cache, though, like queries, or data on azure blob storage, or anything like that.. .

**Reply by TheDoubleH** [reply]
Interesting concept. And While I agree with @NKarolak I also see @waldo1001 s point, in not writing unnecessary code.

That said, I have kept setup tables in memory inside objects many times. But I don't think I have ever done it in a singleinstance codeunit for what You are suggesting.

**Reply by fvet** [reply]
> Ah NST cache, you are right. I will try to find the time to test the performance gain with a single instance codeunit

We've also used the single approach quit often in our legacy code, but did not move forward for new features since the NST caching improved. Curious to see though the performance impact of reading from cache vs access in SingleInstance.

**Comment by StefanMaron** [comment]
Okay, I tested it with this piece of code and the BCPT:

``` AL
trigger OnRun()
var
    SalesSetup: Record "Sales & Receivables Setup";
begin
    if not SalesSetup.Get() then begin
        SalesSetup.Init();
        SalesSetup.Insert();
    end;

    SalesSetup.TestField("Customer Nos.");
    if SalesSetup."Customer Nos." = '' then;
end;
```

It turns out, that even with multiple sessions, the run time is mostly at 0 ms and only a few times at 1 ms. Manual caching does not seem to be needed at all
